### PR TITLE
Week 3 Step 5

### PR DESCRIPTION
**Pull request by ethanimooney**
**TLDR;** Adds Datastore functionality, and ability to display comments on the site itself.
**Link:** https://ethanimooney-step-2020.appspot.com/

This is a long one, sorry...

**pom.xml** - Added the dependency for datastore

**Comment.java** - changed timestamp to a long, added id variable

**DataServlet.java** - renamed as NewCommentServlet.java

**ListCommentsServlet.java** - this servlet lists all of the comments in the Datastore when fetched. It creates a datastore query and then creates an ArrayList of Comment instances, looping through the query and creating an instance and adding it to the ArrayList

**NewCommentServlet.java** - this servlet adds a new comment entity to the datastore when fetched. 

**index.html** - removed placeholder text, added an id for _loadComments()_ to reference, and added the onload call to loadComments() in the body.

**script.js** - deleted old code, added the _loadComments()_ function, which fetches ListCommentServlet.java, loading in the comments from the database. This function then calls the other new function, _createCommentElement()_. It takes in a passed single comment entity from the datastore and creates a commentElement, a div that will be passed back to _loadCommments()_ and then displayed on index.html. It contains the author and message of the comment.

**style.css** - added functionality to make the div where the comments are displayed scrollable once the content becomes too tall for its max height.

### CODE REVIEW

**Comment by ethanimooney** [comment]
Other interns have been commenting on it so don't judge me for the comments haha

**Comment by ethanimooney** [comment]
It's pretty basic in design, will add more tomorrow once I do step 6. :)

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -18,17 +18,23 @@
 
 public final class Comment {
 
-  private final Date timestamp;
+  private final long id;
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:21]: Please add a docstring above describing this class.

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package com.google.sps.servlets;
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:1]: How did you rename this file? It looks like you're losing the version history.

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import com.google.gson.Gson;
+
+//Servlet that creates a new comment entity and adds it to the database
+
+@WebServlet("/new-comment")
+public class NewCommentServlet extends HttpServlet {
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String author = request.getParameter("cf-name");
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:20]: Not sure what 'cf' means in this case.

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import com.google.gson.Gson;
+
+//Servlet that creates a new comment entity and adds it to the database
+
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:16]: Please remove this extra space.

File: portfolio/src/main/webapp/index.html [hunk 5]
```diff
@@ -118,8 +118,7 @@ <h3>Projects</h3>
 							<input type="submit" value="Submit">
               </form>
 					</div>
-					<div class="comments-item">
```
Comment (shaquille-hall) [inline portfolio/src/main/webapp/index.html:121]: Please reformat this file to use 2 spaces for indentations.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -25,8 +25,27 @@ function changeTextBack(text) {
...
+
+  const authorElement = document.createElement('h3');
+  authorElement.innerText = comment.author;
+
+  const messageElement = document.createElement('p');
+  messageElement.innerText = comment.message;
+
+  commentElement.appendChild(authorElement);
+  commentElement.appendChild(messageElement);
+  //taskElement.appendChild(deleteButtonElement);
```
Comment (shaquille-hall) [inline portfolio/src/main/webapp/script.js:49]: If you don't intend on submitting this, please remove.

File: portfolio/src/main/webapp/index.html (hunk 5 above)
Comment (ethanimooney) [reply portfolio/src/main/webapp/index.html:121]: I'm not sure why it looks like that here, on my end on CloudShell it is formatted correctly

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (ethanimooney) [reply portfolio/src/main/webapp/script.js:49]: Got it

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java (hunk 4 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:16]: Done

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java (hunk 3 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:20]: It's the id/name I gave it in the html, "comment-form-name", I'll rename it to that to avoid confusion

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 2 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:1]: I renamed it by right-click in cloudshell. Didn't think about that, any ideas on what to do there? All of the old code was deleted anyway so it is pretty much a new file

**Comment by ethanimooney** [comment]
Updated to reflect suggestions in comments.

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 2 above)
Comment (shaquille-hall) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:1]: git-mv: https://git-scm.com/docs/git-mv

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 1 above)
Comment (shaquille-hall) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:23]: Please reformat to match the standard for multiline javadocs: https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import com.google.sps.data.Comment;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+//Servlet that lists the comments
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:19]: nit: we typically add a space after "//" to improve readability

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      long timestamp = (long) entity.getProperty("timestamp");
+      String author = (String) entity.getProperty("author");
+      String message = (String) entity.getProperty("message");
+
+      Comment comment = new Comment(id, timestamp, author, message);
+      comments.add(comment);
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:38]: This method has significantly lengthened. Perhaps move this logic to a helper. Same for your post method in the other file.

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java (hunk 3 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java]: I had to change these back, it was breaking the entire thing and I can't figure out why. I assume it has something to do with what datastore stores. It was inputting the author and message as blank. I'll work on changing these at another time.

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 7 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:19]: Done

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 8 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:38]: Done

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import com.google.gson.Gson;
+
+//Servlet that creates a new comment entity and adds it to the database
+@WebServlet("/new-comment")
+public class NewCommentServlet extends HttpServlet {
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
[the remark this answers is not in the document]
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:18]: Got it

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 1 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:23]: Got it

**Comment by ethanimooney** [comment]
Committed edits

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 10]
```diff
@@ -13,22 +13,31 @@
 // limitations under the License.
 
 package com.google.sps.data;
-
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:16]: nit: Leaving a space between your package name and your imports is common practice.

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 2 above)
Comment (shaquille-hall) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:1]: It's useful to to go through the practice of reverting changes, which this gives you a nice opportunity to try out. Otherwise, I'd also say it's ok to proceed as is.

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    PreparedQuery results = datastore.prepare(query);
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.setCharacterEncoding("UTF-8"); 
+    response.getWriter().println(gson.toJson(fillList(results)));
+  }
+
+  public List<Comment> fillList(PreparedQuery results){
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:37]: Should this be private? I don't believe it's used elsewhere. Or is it?

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 8 above)
Comment (shaquille-hall) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:46]: Thank you.

File: portfolio/src/main/webapp/index.html (hunk 5 above)
Comment (shaquille-hall) [reply portfolio/src/main/webapp/index.html:121]: Was this updated?

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(createCommentEntity(request));
+    response.sendRedirect("/index.html");
+  }
+
+  public Entity createCommentEntity(HttpServletRequest request) {
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:24]: Should this be private?

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java (hunk 3 above)
Comment (shaquille-hall) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:25]: How was it breaking?

File: portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java (hunk 11 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/ListCommentsServlet.java:37]: Yes should be private, changing

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java (hunk 12 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:24]: Yes should be private, changing

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java (hunk 3 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:25]: Whenever I would submit a new comment with the new id's it would submit to data store with a blank author an message. I assume it has something t do with the id's being stored somewhere I don't know about, because I changed them everywhere in my code

File: portfolio/src/main/webapp/index.html (hunk 5 above)
Comment (ethanimooney) [reply portfolio/src/main/webapp/index.html:121]: I didn't update it no, I really don't know how to make it look different on GitHub. Apparently some other interns were having the issue as well and can't make sense of it either. I have my tab size changed in CloudShell so I am using tabs not spaces

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 10 above)
Comment (ethanimooney) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:16]: Got it

**Comment by ethanimooney** [comment]
Updated

**Comment by ethanimooney** [comment]
Changed the ids for the form, and fixed the issue with submitting to datastore

File: portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+public class NewCommentServlet extends HttpServlet {
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(createCommentEntity(request));
+    response.sendRedirect("/index.html");
+  }
+
+  private Entity createCommentEntity(HttpServletRequest request) {
+    String author = request.getParameter("comment-form-name");
+    String message = request.getParameter("comment-form-message");
```
Comment (shaquille-hall) [inline portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:27]: These are a lot more descriptive! Thanks for bearing through with the changes.

File: portfolio/src/main/webapp/index.html (hunk 5 above)
Comment (shaquille-hall) [reply portfolio/src/main/webapp/index.html:121]: Ack. This was resolved off line and via a new filed issue.
